model-checking/rmc
# fix bit logic operators

- Minimal changes to rustc to enable RMC
- Renamed files to avoid name conflicts
- initial rmc commit
- Fix out-by-one errors in cargo-rmc-tests and run-make tests (#17)
- Add regression test for https://github.com/model-checking/rmc/issues/21 (#22)
- Refactor codegen_vtable and vtable_name (#24)
- Correctly track the type of boxed fat pointers when projecting through them (#23)
-  Fix issue where array types were recorded as [x; _] and hence aliasing (#15)
- Conform with latest changes to `Binder` (used in closures) (#27)
- Squashed a bunch of warnings (#28)
- Squash warnings: remove final use of insert_unchecked (#29)
- Properly handle Fat-ptr to Fat-ptr casts (#62)
- Add RMC workflow file for GitHub Actions (#31)
- Add vtable well-formedness flag
- Add templates for Pull Requests and Issues (#81)
- Add a couple of missing tests (nondet. vectors, offset) (#79)
- Complete and replace the unsized pointer cast implementation (#88)
- Add script and workflow for copyright checks in CI (#92)
- Add unsized pointer cast regression tests (#89)
- update readme (#91)
- Add nondet. bytes test (#94)
- Enable regression workflow for macOS (#97)
- Allow cast of any pointer to cast to a trait pointer (#98)
- Allow cast of fat pointers to thin pointers (#99)
- Minor tweaks to the documentation (#100)
- Correct fat pointer generation by codegen_rvalue_ref (#101)
- Add box mut dyn trait regression test (#108)
- Avoid strict type comparison in statement assignment (#110)
- Extend `entry_fn` call to use local defid
- Implemented __VERIFIER_assume hook (#115)
- Conform with latest changes in crate handling
- Disable regression test failing in CI but not locally (#133)
- Refactor: cleanup hooks.rs (#116)
- Use num::BigInt instead of i128 in constants, as constants can be of arbitrary size. (#141)
- Added Location to stmt constructors(#139)
- Use canonical filenames in source locations (#142)
- Relax the requirements for is_rust_box() to enable crate compilation (#153)
- Correctly handle the abort intrinsic (#151)
- Improve some regression tests and add some more. (#154)
- Added argument casting when type is same. (#137)
- Complete count intrinsics codegen and testing (#155)
- Install CBMC viewer in CI (#162)
- Added fixme tests for 'Unable to find vtable symbol' bug. (#159)
- Import and use `LOCAL_CRATE` in `crate_name` call
- Update CBMC version check to 5.30.0 (fix #164) #166
- Move 1 rust-test to compiletest. (#163)
- Move balloon standalone example to firecracker folder (#167)
- Refactor pretty name to restore passable dyn trait functionality (#171)
- Added v0 mangler as default for rmc, and added --mangler option. (#170)
- Added --visualize flag to rmc. (#138)
- Move all tests under rust-tests to compiletest. (#168)
- Fix for duplicate vtable field names (#188)
- Add --gen-c flag to cargo-rmc. (#192)
- Remove slow find_by_pretty_name vtable search (#193)
- Refactor: Cleanly seperate `CurrentFnCtx` from `GotocCtx`. (#195)
- Added check for cbmc-viewer version. (#174)
- Add runtime asserts and tests for vtable size/align (#197)
- Move cargo tests to compiletest. (#198)
- Move now-working boxed dyn trait test (#211)
- Initial changes required to codegen stdlib (#210)
- Separate RMC sanity check function (#201)
- Conform with refactor of vtable codegen
- Move gotoc tests under run-make to their own test suite. (#217)
- Add rmc-prelude.rs, and use it in tests (#226)
- Use RMC prelude in all tests (#232)
- Add an __VERIFIER_expected_fail() intrinsic, and use it in a test (#229)
- Codegen drop_in_place type and vtable pointer (#222)
- Handle copy on empty str/length zero data (#235)
- Enforce formatting of RMC tests. (#258)
- Added --function flag to cargo-rmc (#239)
-  Handle unsized dyn trait objects as source of casts (#233)
- Use a better binder/substs for vtable instance lookup.  (#257)
- Implement sizeof for Code for vtable generation (fix crash) (#261)
- Fix types for codegen of intrinsic const (#268)
- Support atomic exchange primitives (#266)
- Refactor: Use borrowed types for arguments (#269)
- use codegen_unimplemented() for the try intrinsic (#275)
- Correct check for vtable pointer type (fix Firecracker regression) (#279)
- Handle boxed closures as arguments (fix RMC std lib crash) (#278)
- Add support for multiple expected failures. (#280)
-  Codegen the rest of std lib and add a regression script
- Update dyn trait tests from raw to ptr Rust APIs (#306)
- Custom panic hook for RMC including current item (#304)
- Conform with recent changes in rustc (limits, binders) (#318)
- Fix codegen for dynamic `Fn::call` and `FnMut::call_mut` (#320)
- Use default CBMC flags in RMC (#263)
- Improve setup scripts for development (#322)
- C lib flag (#260)
- Emit an error message when rmc-rustc fails (#328)
- Codegen new intrinsic `raw_eq` (#331)
- Use Rust's vtable_entries (#329)
- Extract examples from reference and display their results in a dashboard. (#325)
- Symbol table transformer (#321)
- Use `x.py` to build and run the dashboard. (#333)
- Fix dyn trait _fail tests compilation errors (#342)
- Remove string that tripped open source alarm. (#344)
- Differentiate between check, codegen, and verification errors. (#353)
- Generics in dyn trait fix: use vtable index to resolve functions (#352)
- Add support for simd insert/extract intrinsics (#354)
- Conform with recent changes in rustc (alloc id, scalar ptr, vtable method) (#361)
- Cleanup rmc flags (#365)
- Fix dry-run to print to stdout when output_to is not stdout. (#366)
- Detect which reference examples should check/codegen/verify. (#363)
- Correct function type for vtable shim (fix some boxed closures) (#335)
- Support cargo rmc flags (#371)
- Removed incorrect special case for Rust bitwise operators


### Description of changes: 

Describe RMC's current behavior and how your code changes that behavior. If there are no issues this PR is resolving, explain why this change is necessary.

### Resolved issues:

Resolves #ISSUE-NUMBER


### Call-outs:



### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.
